Modules/_tkinter.c

```diff
@@ -3040,10 +3040,15 @@ static void
 Tkapp_Dealloc(PyObject *self)
 {
     PyObject *tp = (PyObject *) Py_TYPE(self);
-    /*CHECK_TCL_APPARTMENT;*/
-    ENTER_TCL
-    Tcl_DeleteInterp(Tkapp_Interp(self));
-    LEAVE_TCL
+    if (((TkappObject *)self)->threaded && \
+        ((TkappObject *)self)->thread_id != Tcl_GetCurrentThread()) { \
+        // We cannot delete the interpreter in the wrong thread (bpo-39093)
+        PyErr_WriteUnraisable(PyUnicode_FromString("Deallocation of Tkapp attempted in wrong thread. Skipping deletion of Tcl interpreter (this will cause a memory leak).")); \
+    } else {
```

Comment:
Did you avoid using the CHECK_TCL_APPARTMENT macro to get a chance to use PyErr_WriteUnraisable? Might be worth a comment.

Reply:
I am not sure about a comment, as it is hopefully obvious because this is the main use-case for [`PyErr_WriteUnraisable`](https://docs.python.org/3/c-api/exceptions.html#c.PyErr_WriteUnraisable). I needed to avoid `CHECK_TCL_APPARTMENT` because that would raise a RuntimeError which, in-turn, would cause a SystemError to be raised (and, as this is **not** a bug with the Python interpreter, we need to avoid that)

Reply:
Regarding the interpreter crash, this PR seems to fix it. Linking tqdm/tqdm#1007 for the record. Note that the long lived refs are still a bug in tqdm, this PR makes them just a very small memory leak instead of a catastrophic crash.